cpp/src/neighbors/detail/vamana/vamana_serialize.cuh

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "vamana_structs.cuh"
+
+#include <cuvs/neighbors/vamana.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/logger-ext.hpp>
+#include <raft/core/mdarray.hpp>
+#include <raft/core/mdspan_types.hpp>
+#include <raft/core/nvtx.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/serialize.hpp>
+
+#include "../dataset_serialize.hpp"
+
+#include <cstddef>
+#include <cstdint>
+#include <fstream>
+#include <type_traits>
+
+namespace cuvs::neighbors::vamana::detail {
+
+/**
+ * Save the index to file.
+ *
+ * Experimental, both the API and the serialization format are subject to change.
+ *
+ * @param[in] res the raft resource handle
+ * @param[in] file_name the path and name of the DiskAN index file generated
+ * @param[in] index_ VAMANA index
+ *
+ */
+
+template <typename T, typename IdxT>
+void serialize(raft::resources const& res,
+               const std::string& file_name,
+               const index<T, IdxT>& index_,
+               bool include_dataset)
+{
```

Comment:
I believe the medoid also needs to be saved to begin greedy search from the same point later.

Reply:
Need to confirm this with the DiskANN source code though.

Reply:
Oh, right, good catch! Will add this.

Reply:
Fixed this, medoid is now saved in the index structure and correctly output with `serialize`